app/src/main/java/com/github/se/eduverse/ui/search/UserProfileScreen.kt

```diff
@@ -125,13 +129,41 @@ fun UserProfileScreen(
                       }
 
                   // Follow/Unfollow Button
-                  val currentUserId = FirebaseAuth.getInstance().currentUser?.uid
                   if (currentUserId != null && currentUserId != userId) {
                     Button(
-                        onClick = { TODO() },
-                        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp)) {
-                          Text(if (profile.followers > 0) "Unfollow" else "Follow")
+                        onClick = { viewModel.toggleFollow(currentUserId, userId) },
+                        modifier =
+                            Modifier.fillMaxWidth()
+                                .padding(horizontal = 16.dp)
+                                .testTag("follow_button"),
+                        enabled = followActionState != FollowActionState.Loading) {
+                          when (followActionState) {
+                            FollowActionState.Loading -> {
+                              CircularProgressIndicator(
+                                  modifier = Modifier.size(24.dp),
+                                  color = MaterialTheme.colorScheme.onPrimary)
+                            }
```

Comment:
In the followActionState state management, it is a good idea to clearly indicate a user message or an appropriate action in case of loading or failure of the action. However, for the FollowActionState.Loading state, displaying the CircularProgressIndicator is fine, but consider adding a ContentDescription for accessibility.